src/middle/Mir_pattern.ml

```diff
@@ -0,0 +1,538 @@
+end
+
+module Make_traversable_outvar2(A:Applicative.S2) 
+  : Traversable.S2 with module A := A and type 'a t := 'a outvar 
+= struct
+  module Traversable_sizedtype = Make_traversable_sizedtype2(A)
+
+  let traverse ({out_constrained_st;out_unconstrained_st;_} as outvar) ~f = 
+    A.map2 
+      ~f:(fun out_constrained_st out_unconstrained_st -> 
+          {outvar with 
+            out_constrained_st = out_constrained_st
+            ; out_unconstrained_st = out_unconstrained_st
+           }
+      ) 
+      (Traversable_sizedtype.traverse ~f out_constrained_st)
+      (Traversable_sizedtype.traverse ~f out_unconstrained_st)
+end
+
+type ('e, 's) prog =
+  { functions_block: 's fun_def list
+  ; input_vars: (string * 'e sizedtype) list
+  ; prepare_data: 's list (* data & transformed data decls and statements *)
+  ; log_prob: 's list (*assumes data & params are in scope and ready*)
+  ; generate_quantities: 's list (* assumes data & params ready & in scope*)
+  ; transform_inits: 's list
+  ; output_vars: (string * 'e outvar) list
+  ; prog_name: string
+  ; prog_path: string }
+[@@deriving sexp, map]
+
+
+module Make_traversable_prog(A:Applicative.S) 
+  : Bitraversable.S with module A := A and type ('a,'b) t := ('a,'b) prog
+= struct
+
+  module Traversable_fun_def = Make_traversable_fun_def(A)
+  module Traversable_sizedtype = Make_traversable_sizedtype(A)
+  module Traversable_outvar = Make_traversable_outvar(A)
+
+  let traverse prog ~f ~g = 
+    let functions_block = 
+      prog.functions_block
+      |> List.map ~f:(Traversable_fun_def.traverse ~f:g)   
+      |> A.all
+    and input_vars = 
+      prog.input_vars
+      |> List.map ~f:(fun (str,st) -> 
+          Traversable_sizedtype.traverse ~f st 
+          |> A.map ~f:(fun st -> (str,st))
+          ) 
+      |> A.all
+    and prepare_data = 
+      List.map prog.prepare_data ~f:g
+      |> A.all 
+    and log_prob = 
+      List.map prog.log_prob ~f:g
+      |> A.all 
+    and transform_inits = 
+      List.map prog.transform_inits ~f:g
+      |> A.all 
+    and generate_quantities = 
+      List.map prog.generate_quantities ~f:g
+      |> A.all 
+    and output_vars = 
+      List.map prog.output_vars 
+        ~f:(fun (str,outvar) -> 
+          Traversable_outvar.traverse ~f outvar 
+          |> A.map ~f:(fun st -> (str,st))) 
+      |> A.all
+    in
+
+    A.(return (fun functions_block input_vars prepare_data log_prob generate_quantities transform_inits output_vars  ->
+      { prog with 
+        functions_block = functions_block
+      ; input_vars = input_vars
+      ; prepare_data = prepare_data 
+      ; log_prob = log_prob 
+      ; generate_quantities = generate_quantities 
+      ; transform_inits = transform_inits
+      ; output_vars = output_vars  
+      }
+    )
+    <*> functions_block
+    <*> input_vars
+    <*> prepare_data
+    <*> log_prob
+    <*> generate_quantities
+    <*> transform_inits
+    <*> output_vars
```

Comment:
This style is so so so much more verbose and hard to understand for me than the simple version of this would be.

Reply:
Hmm.. what did you have in mind - I'm not quite sure what the simple version is here. 

Note that each of these values is lifted into the applicative (i.e. `transform_inits` has type `'s list A.t`) OCaml doesn't have a `map7`. 

We also can't use `bind` since traverse is only defined on applicative rather than monad.

Reply:
I mean like a simple traverse implementation with signature something like `ref 'a -> ('b -> 'c) -> 'b -> 'c`.

Reply:
It also looks like we might be able to use `map` + `iter`: https://github.com/ocaml-ppx/ppx_deriving#plugins-iter-map-and-fold

Reply:
You could definitely do that for a traverse where the effect was state (in the same way as you could do a map with a ref cell rather than a fold) but what about where the applicative is something else e.g. validation? I'd also say when the effect in  traverse is state map_accum_left / right (defined in Program.ml) provide a clean interface.